mlflow/cli.py

```diff
@@ -180,6 +181,7 @@ def server(file_store, default_artifact_root, host, port, workers):
 cli.add_command(mlflow.sagemaker.cli.commands)
 cli.add_command(mlflow.azureml.cli.commands)
 cli.add_command(mlflow.experiments.commands)
+cli.add_command(mlflow.utils.internal_cli.commands)
```

Comment:
We use a new CLI command instead of invoking `mlflow run` directly because:
* There's some setup in `mlflow run` that should happen just-once in an environment (e.g. multiple concurrent attempts to create the same conda env can fail)
* I didn't want `mlflow run` to have different behavior when invoked by the user in a shell vs in `mlflow.projects.run()`. Without a separate CLI command, we'd need to have `mlflow run` conditionally reuse an existing run ID / log parameters depending on whether it was being invoked by a user or in `mlflow.projects.run`

Reply:
Could we do this through a flag on `mlflow run` instead, similar to how we have no-conda and stuff like that? If not, is there any way to hide this from the help for `mlflow`? It's really weird to see an internal command there.

Reply:
Also, I didn't quite understand which bits of `mlflow run` we don't want in this case. For creating Conda environments, does this solution somehow prevent people from running `mlflow run` multiple times on their own, or spawning multiple async runs? If not, it seems like we should make that recover from the concurrent Conda issue. For the run ID part, maybe `mlflow run` can take an optional run ID to use, and not start its own when this special flag is passed (as mentioned above).

Reply:
Looks like maybe `hidden=True` can hide commands in Click: https://github.com/pallets/click/pull/500

Reply:
Agree that it's weird to see an internal command, thanks for tip RE hiding commands. Regarding your questions:

> Could we do this through a flag on mlflow run instead, similar to how we have no-conda and stuff like that?

We could have `mlflow run`'s behavior vary via a flag (or by adding a new undocumented 'mode' e.g. `mlflow run -m run_entry_point ...`).

It's tricky to have `mlflow run` invoke itself as there's a set `S` of things we want to happen just-once during a single project run. For example, across (a) an invocation of `mlflow run` launched by the user via the CLI and (b) one launched via `subprocess.Popen` in `mlflow.projects.run` we want to create a new run and log parameters just once.

We could get around this by passing different arguments (i.e. a run ID when we don't want to create a run) / using `mode` to determine the right behavior, my concern is if/as `S` grows, we have to keep adding arguments to `mlflow run` (& mark these arguments as internal-use only). By running separate commands we can more easily decouple cases (a) and (b).

However I'm open to running everything via `mlflow run`, don't have a super strong opinion either way.

> Also, I didn't quite understand which bits of mlflow run we don't want in this case. For creating Conda environments, does this solution somehow prevent people from running mlflow run multiple times on their own, or spawning multiple async runs? If not, it seems like we should make that recover from the concurrent Conda issue. 

With this PR users can spawn multiple async runs of the same project via the Python API; the conda environment will be created synchronously for the first run of the project, so parallel conda env creation shouldn't be a problem.

This PR doesn't prevent people from invoking `mlflow run` concurrently e.g. in separate terminal windows (conda env creation would break here), although hopefully this use case isn't needed if users can launch concurrent runs via the Python API. 

> Looks like maybe hidden=True can hide commands in Click: pallets/click#500

Thanks for the tip! Unfortunately seems this functionality will only be released with Click 7.0 / it's not currently available

Reply:
Ah, I see. In that case maybe a special "mode" is the best way. For example `local-noenv` or something.